### ISSUE TITLE
Support `switch` statements

### ISSUE THREAD

**Issue by es128** [eventualbuddha/decaffeinate#50]
Trying out decaffeinate, unfortunately not getting very far with it. Didn't notice an existing issue for the lack of `switch` support, which I bumped into quickly, so figured I'd open one.

```sh
λ decaffeinate src/test.coffee
[path-to]/decaffeinate/lib/utils/traverse.js:142
    throw new Error('cannot traverse unknown node type: ' + node.type);
          ^
Error: cannot traverse unknown node type: Switch
```

**Comment by eventualbuddha** [comment]
Thanks @es128. I added support for `switch` statements just now. Can you try v1.28.0 and let me know here if you encounter any other issues with `switch`? If you find other problems unrelated to `switch` please file another issue.

**Comment by es128** [comment]
It's not failing as before, but implicit `return` is not working properly. For blocks like this that don't appear to have any ES6 advantages or otherwise aren't handled in decaffeinate, why not have a generic handler that just uses the default coffeescript compilation?

I am still seeing other obfuscated errors in my use-case, but don't have the time atm to fully trace what's going on. Feel free to dig into it if you like, I was trying it on https://github.com/es128/progeny/blob/master/src/index.coffee.

```sh
Error: Syntax error on line 1, column 1: unexpected '=' (\u003D)
1 : 'use strict'
^ : ^
```

There is some other more substantial non-public code I was hoping to use this with - as you make updates I'd be happy to keep trying it and giving you feedback about what I'm seeing if you like.

Either way, this is an interesting project - thank you for the work you've put into it.

**Comment by eventualbuddha** [comment]
> implicit return is not working properly

This is now handled correctly:

```coffee
->
  switch a
    when b
      c
```

```js
(function() {
  switch (a) {
    case b:
      return c;
  }
});
```

> For blocks like this that don't appear to have any ES6 advantages or otherwise aren't handled in decaffeinate, why not have a generic handler that just uses the default coffeescript compilation?

A big motivator for this project over simply using the default CoffeeScript compilation is that when you pass your source code through `coffee` the output is stripped of all comments and formatting. If the goal is to convert your CoffeeScript code to JavaScript code permanently, this is unacceptable.

> I am still seeing other obfuscated errors in my use-case, but don't have the time atm to fully trace what's going on.

That's okay, pointing me to files that fail in some way is still useful. For the file you pointed out, based on the error message you gave, it appears that CoffeeScriptRedux is failing to parse it, meaning there's some difference between how CoffeeScript handles the code and CSR handles it. Unfortunately CSR's error messages are not always the best and sometimes simply point to the first character in the file. The `unexpected '='` should give a clue as to where it's failing, though.

Please keep the issues coming  :smiley:

**Comment by eventualbuddha** [comment]
I just added support for the last two missing `switch` features I could find:

### `switch` without an expression

```coffee
switch
  when score < 60 then 'F'
  when score < 70 then 'D'
  when score < 80 then 'C'
  when score < 90 then 'B'
  else 'A'
```

```js
switch (false) {
  case !(score < 60): 'F'; break;
  case !(score < 70): 'D'; break;
  case !(score < 80): 'C'; break;
  case !(score < 90): 'B'; break;
  default: 'A';
}
```

### `switch` used as an expression

```coffee
a = switch b
  when c then d
  else e
```

```js
var a = (function() {
  switch (b) {
    case c: return d;
    default: return e;
  }
})();
```

**Comment by es128** [comment]
Cool. Although one minor thing I've noticed is some odd indentation at the end of the IIFE wrapping. When I try your second example it renders out like

```js
var a = (function() {
  switch (b) {
    case c: return d;
    default: return e;
  }
}
  )();
```

**Comment by eventualbuddha** [comment]
That's with v1.29.1? Can you let me know what exactly you're running to get that result?

**Comment by es128** [comment]
Yes, and something similar was happening with earlier versions too. Command is just `decaffeinate src/test.coffee` after copying your example into that file.

```
λ cat ~/.nvm/versions/io.js/v2.4.0/lib/node_modules/decaffeinate/package.json
{
  "name": "decaffeinate",
  "version": "1.29.1",
```

**Comment by eventualbuddha** [comment]
Found it. Turns out it was to do with whether or not there was additional newlines at the end of the file. Here's the fix: 14655f7529271c980bfe1a2ac458a82b48569a81. Released in v1.29.2.

**Comment by es128** [comment]
Cool, thanks. Been trying it out some more, but now realizing that I'm running into all sorts of issues that come down to previously reported issues with redux's parsing. I had no idea, but guess I was doing some [adventurous things](https://github.com/es128/progeny/blob/master/src/index.coffee#L45-L54) with my coffeescript syntax that redux really chokes all over.

Will keep playing with it, and will report anything I find to be specific to decaffeinate.
